src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts

```diff
@@ -291,6 +401,212 @@ export async function putDataInTextureThenDrawAndCheckResults<T extends Dimensio
   return errs.length > 0 ? new Error(errs.join('\n')) : undefined;
 }
 
+/**
+ * "Renders a quad" to a TexelView with the given parameters,
+ * sampling from the given Texture.
+ */
+export function softwareRasterize<T extends Dimensionality>(
+  texture: Texture,
+  sampler: GPUSamplerDescriptor,
+  targetSize: [number, number],
+  options: TextureTestOptions
+) {
+  const [width, height] = targetSize;
+  const { ddx = 1, ddy = 1, ddxStart = 0, ddyStart = 0 } = options;
+  const format = 'rgba32float';
+
+  const rep = kTexelRepresentationInfo[format];
+
+  const expData = new Float32Array(width * height * 4);
+  for (let y = 0; y < height; ++y) {
+    const fragY = height - y - 1 + 0.5;
+    for (let x = 0; x < width; ++x) {
+      const fragX = x + 0.5;
+      const coords = [(fragX / width) * ddx - ddxStart, (fragY / height) * ddy - ddyStart] as T;
```

Comment:
This bit of math is the most mysterious to me. Comment it?

Reply:
added

Reply:
Ahh, totally missed that this was just a copy of the same math you did in the WGSL. Explains why it felt sort of arbitrary :)